rewrite-gradle/src/main/java/org/openrewrite/gradle/UpgradeDependencyVersion.java

```diff
@@ -116,189 +123,275 @@ public Validated<Object> validate() {
-                    } else if (arg instanceof J.Literal) {
-                        J.Literal literal = (J.Literal) arg;
-                        String gav = (String) literal.getValue();
-                        if (gav == null) {
-                            getCursor().putMessage(UPDATE_VERSION_ERROR_KEY, new IllegalStateException("Unable to update version"));
-                            return arg;
+            @Override
+            public J postVisit(J tree, ExecutionContext ctx) {
+                if (tree instanceof JavaSourceFile) {
+                    JavaSourceFile cu = (JavaSourceFile) tree;
+                    Optional<GradleProject> maybeGp = cu.getMarkers().findFirst(GradleProject.class);
+                    maybeGp.ifPresent(gradleProject -> acc.gradleProject = gradleProject);
+                }
+                return tree;
+            }
+        };
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor(DependencyVersionState acc) {
+        MethodMatcher dependencyDsl = new MethodMatcher("DependencyHandlerSpec *(..)");
+        DependencyMatcher dependencyMatcher = new DependencyMatcher(groupId, artifactId, null);
+        return
+            Preconditions.or(
+                new PropertiesVisitor<ExecutionContext>() {
+                    @Override
+                    public boolean isAcceptable(SourceFile sourceFile, ExecutionContext ctx) {
+                        if (!super.isAcceptable(sourceFile, ctx)) {
+                            return false;
                         }
-                        Dependency dep = DependencyStringNotationConverter.parse(gav);
-                        if (dependencyMatcher.matches(dep.getGroupId(), dep.getArtifactId())
-                            && dep.getVersion() != null
-                            && !dep.getVersion().startsWith("$")) {
-                            GradleProject gradleProject = getCursor().firstEnclosingOrThrow(JavaSourceFile.class)
-                                    .getMarkers()
-                                    .findFirst(GradleProject.class)
-                                    .orElseThrow(() -> new IllegalArgumentException("Gradle files are expected to have a GradleProject marker."));
-                            String version = dep.getVersion();
-                            try {
-                                String newVersion = "classpath".equals(method.getSimpleName()) ?
-                                        findNewerPluginVersion(dep.getGroupId(), dep.getArtifactId(), version, gradleProject, ctx) :
-                                        findNewerProjectDependencyVersion(dep.getGroupId(), dep.getArtifactId(), version, gradleProject, ctx);
-                                if (newVersion == null || version.equals(newVersion)) {
-                                    return arg;
+                        return equalIgnoringSeparators(sourceFile.getSourcePath(), Paths.get(GRADLE_PROPERTIES_FILE_NAME));
```

Comment:
We will either want to...

* Use a `Preconditions.check(new FindSourceFiles("**/gradle.properties"), <existing visitor>)`
* Update this comparator to allow matching for any file with a name `gradle.properties` as they can appear multiple times throughout a Gradle multi module project.

Reply:
@shanman190 what about something like this which is used in AddProperty recipe?
`sourceFile.getSourcePath().endsWith("gradle.properties"))`

Reply:
@ajohnsonz, yep! That fits with option 2 that I mentioned previously.